src/annoylib.h

```diff
@@ -940,23 +963,32 @@ template<typename S, typename T, typename Distance, typename Random>
       return false;
     }
 
+    if (n_threads < 1) {
+      set_error_from_string(error, "You can't build an index with less than 1 thread");
+      return false;
+    }
+
     D::template preprocess<T, S, Node>(_nodes, _s, _n_items, _f);
 
     _n_nodes = _n_items;
-    while (1) {
-      if (q == -1 && _n_nodes >= _n_items * 2)
-        break;
-      if (q != -1 && _roots.size() >= (size_t)q)
-        break;
-      if (_verbose) showUpdate("pass %zd...\n", _roots.size());
 
-      vector<S> indices;
-      for (S i = 0; i < _n_items; i++) {
-        if (_get(i)->n_descendants >= 1) // Issue #223
-          indices.push_back(i);
+    std::mutex _nodes_mutex;
+    ThreadBarrier barrier(n_threads);
+    vector<std::thread> threads(n_threads);
+    int work_per_thread = (int)floor(q / (double)n_threads);
+    int work_remainder = q % n_threads;
+    for (int i = 0; i < n_threads; i++) {
+      int trees_per_thread = -1;
+      if (q > -1) {
+        // First thread picks up the remainder of the work
```

Comment:
this might lead to an uneven split, eg if `q=20` and `n_threads=7` then the distribution will be `[8, 2, 2, 2, 2, 2, 2]`

I think you can do `trees_per_thread  work_per_thread + (i < work_remainder)` and it should split it like `[3, 3, 3, 3, 3, 3, 2]`

Reply:
actually even simpler you can just to `trees_per_thread = (q + i) / n_threads` and you don't even have to define `work_pre_thread`

Reply:
Awesome, I was not exactly happy with my solution, so I'll reuse this.